File: kotlin-app/app/src/main/java/io/github/kawase/socket/packet/Packet.java
package io.github.kawase.socket.packet;

import io.github.kawase.socket.exceptions.PacketException;
import io.github.kawase.socket.interfaces.Data;
import io.github.kawase.socket.utility.EncryptionUtility;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

/**
 * The parent packet class which every packet will extend
 * it uses AES to encrypt and decrypt and uses a basic dynamic key system
 * <p>
 * additional notes:
 * We could only write to the byte buffer encrypted data every time instead of encrypting it when sending it.
 * We could use a better encryption algorithm like ChaCha20 with poly 1305 and mac
 * We could also use Diffie-Hellman + a better dynamic key system for more security,
 * but none of that is rlly needed for a socket with the sole purpose of being used for an IRC.
 */
public abstract class Packet {
    private final int id;

    public Packet(final int id) {
        this.id = id;
    }

    public int getId() {
        return id;
    }

    /**
     * The children classes will override these methods to write / read their data into their fields.
     */
    protected abstract void write(final ByteBuffer buffer);
    protected abstract void read(final ByteBuffer buffer);

    public ByteBuffer encode() {
        try {
            final long dynamicSeed = System.nanoTime();

            final byte[] encryptedSeed = EncryptionUtility.encryptLong(dynamicSeed, Data.baseKey);
            final ByteBuffer payloadBuffer = ByteBuffer.allocate(4096);

            payloadBuffer.putInt(id);
            this.write(payloadBuffer);
            payloadBuffer.flip();

            final byte[] payloadBytes = new byte[payloadBuffer.remaining()];
            payloadBuffer.get(payloadBytes);

            final byte[] encryptedPayload = EncryptionUtility.encryptBytes(payloadBytes, String.valueOf(dynamicSeed));
            final ByteBuffer finalBuffer = ByteBuffer.allocate(Integer.BYTES + encryptedSeed.length + encryptedPayload.length);

            finalBuffer.putInt(encryptedSeed.length);
            finalBuffer.put(encryptedSeed);
            finalBuffer.put(encryptedPayload);

            finalBuffer.flip();

            return finalBuffer;
        } catch (Exception e) {
            throw new PacketException("Failed to encrypt", e.getCause());
        }
    }

    public void decode(final ByteBuffer byteBuffer) {
        this.read(byteBuffer);
    }

    /**
     * Constructs a packet object with the unencrypted data in it.
     * @param byteBuffer the buffer where the encrypted data is stored.
     * @param packetManager I wonder
     * @return the packet instance with the data in it
     * @throws Exception if the decryption fails.
     */
    public static Packet construct(final ByteBuffer byteBuffer, final PacketManager packetManager) throws Exception {
        final int seedLength = byteBuffer.getInt();

        // this should prevent out of memory attacks.
        if (seedLength <= 0 || seedLength > 1024) {
            throw new PacketException("Invalid seed length");
        }

        final byte[] encryptedSeed = new byte[seedLength];
        byteBuffer.get(encryptedSeed);

        final long dynamicSeed = EncryptionUtility.decryptLong(encryptedSeed, Data.baseKey);

        final byte[] encryptedPayload = new byte[byteBuffer.remaining()];
        byteBuffer.get(encryptedPayload);

        final byte[] decryptedPayloadBytes = EncryptionUtility.decryptBytes(encryptedPayload, String.valueOf(dynamicSeed));
        final ByteBuffer decryptedBuffer = ByteBuffer.wrap(decryptedPayloadBytes);

        final int packetID = decryptedBuffer.getInt();

        final Packet packet = packetManager.createPacket(packetID);
        packet.decode(decryptedBuffer);

        return packet;
    }

    /**
     * Helper for writing a string into a byte buffer.
     * @param data data to write
     * @param buffer the buffer to write the data to
     */
    public void putString(final String data, final ByteBuffer buffer) {
        final byte[] stringBytes = data.getBytes(StandardCharsets.UTF_8);

        buffer.putInt(stringBytes.length);
        buffer.put(stringBytes);
    }

    /**
     * Helper for reading a string from a buffer.
     * @param buffer the buffer to read from
     * @return the string it read from the buffer
     */
    public String readString(final ByteBuffer buffer) {
        final int length = buffer.getInt();
        final byte[] bytes = new byte[length];

        buffer.get(bytes);

        return new String(bytes, StandardCharsets.UTF_8);
    }
}

File: unity/NeuroKey/Assets/Scripts/Runtime/Network/Packet.cs
using System;
using System.IO;
using System.Text;

namespace NeuroKey.Network
{
    public abstract class Packet
    {
        public const string BaseKey = "CIOCLIKESKIDSIJIJSDJ1J2313J8123869699696";
        private readonly int id;

        protected Packet(int id)
        {
            this.id = id;
        }

        public int Id => id;

        protected abstract void Write(BinaryWriter writer);
        protected abstract void Read(BinaryReader reader);

        public byte[] Encode()
        {
            long dynamicSeed = (long)(DateTime.UtcNow.Subtract(new DateTime(1970, 1, 1))).TotalMilliseconds * 1000000; // NanoTime simulation
            byte[] encryptedSeed = EncryptionUtility.EncryptLong(dynamicSeed, BaseKey);

            using (var payloadStream = new MemoryStream())
            using (var writer = new BinaryWriter(payloadStream))
            {
                // Write ID as BigEndian Int
                WriteInt32BigEndian(writer, id);
                Write(writer);
                writer.Flush();
                byte[] payloadBytes = payloadStream.ToArray();

                byte[] encryptedPayload = EncryptionUtility.EncryptBytes(payloadBytes, dynamicSeed.ToString());

                using (var finalStream = new MemoryStream())
                using (var finalWriter = new BinaryWriter(finalStream))
                {
                    // Write seed length as BigEndian
                    byte[] seedLengthBytes = IntToBigEndian(encryptedSeed.Length);
                    finalWriter.Write(seedLengthBytes);
                    finalWriter.Write(encryptedSeed);
                    finalWriter.Write(encryptedPayload);
                    return finalStream.ToArray();
                }
            }
        }

        public static Packet Decode(byte[] bytes, PacketManager manager)
        {
            int seedLength = BigEndianToInt(bytes, 0);
            byte[] encryptedSeed = new byte[seedLength];
            Buffer.BlockCopy(bytes, 4, encryptedSeed, 0, seedLength);

            long dynamicSeed = EncryptionUtility.DecryptLong(encryptedSeed, BaseKey);

            int payloadOffset = 4 + seedLength;
            int payloadLength = bytes.Length - payloadOffset;
            byte[] encryptedPayload = new byte[payloadLength];
            Buffer.BlockCopy(bytes, payloadOffset, encryptedPayload, 0, payloadLength);

            byte[] decryptedPayload = EncryptionUtility.DecryptBytes(encryptedPayload, dynamicSeed.ToString());

            using (var stream = new MemoryStream(decryptedPayload))
            using (var reader = new BinaryReader(stream))
            {
                int packetId = BigEndianToInt(decryptedPayload, 0);
                stream.Position = 4;
                
                Packet packet = manager.CreatePacket(packetId);
                packet.Read(reader);
                return packet;
            }
        }

        protected void PutString(BinaryWriter writer, string data)
        {
            byte[] bytes = Encoding.UTF8.GetBytes(data);
            WriteInt32BigEndian(writer, bytes.Length);
            writer.Write(bytes);
        }

        protected string ReadString(BinaryReader reader)
        {
            int length = ReadInt32BigEndian(reader);
            byte[] bytes = reader.ReadBytes(length);
            return Encoding.UTF8.GetString(bytes);
        }

        // Helper methods for BigEndian (since Java uses it)
        public static void WriteInt32BigEndian(BinaryWriter writer, int value)
        {
            byte[] bytes = BitConverter.GetBytes(value);
            if (BitConverter.IsLittleEndian) Array.Reverse(bytes);
            writer.Write(bytes);
        }

        public static int ReadInt32BigEndian(BinaryReader reader)
        {
            byte[] bytes = reader.ReadBytes(4);
            if (BitConverter.IsLittleEndian) Array.Reverse(bytes);
            return BitConverter.ToInt32(bytes, 0);
        }

        public static byte[] IntToBigEndian(int value)
        {
            byte[] bytes = BitConverter.GetBytes(value);
            if (BitConverter.IsLittleEndian) Array.Reverse(bytes);
            return bytes;
        }

        public static int BigEndianToInt(byte[] bytes, int offset)
        {
            byte[] data = new byte[4];
            Buffer.BlockCopy(bytes, offset, data, 0, 4);
            if (BitConverter.IsLittleEndian) Array.Reverse(data);
            return BitConverter.ToInt32(data, 0);
        }
    }
}
